Initial run of role

greg-hellings opened the pull request:
From

greg-hellings committed:
Initial run of role

robled commented on defaults/main.yml:
```diff
@@ -0,0 +1,42 @@
+passwordless_ssh_user: root
+# The home directory for the user who is being configured.
+passwordless_ssh_user_home: /root
+# Name that the prviate key file will be uploaded to. This should match the
```
typo - _private_

robled commented on defaults/main.yml:
```diff
@@ -0,0 +1,42 @@
+# over the user directory where SSH is being configured
+passwordless_ssh_become: true
+# The user to sudo/become. If this gets changed from being the same as the user
+# then unexpected thigns can happen when setting up the known_hosts, as the
```
typo - _things_

andrew103 commented on README.md:
```diff
@@ -0,0 +1,77 @@
+      passwordless_ssh_private_key: "{{ lookup('env', 'HOME') }}/mytestkeys/id_rsa"
+      passwordless_ssh_public_key: "{{ lookup('env', 'HOME') }}/mytestkeys/id_rsa.pub"
+      passwordless_ssh_user: testuser
+      passwordless_ssh_user_dir: /home/testuser
```
@seandst @robled  Would it be considered better practice to declare the variables under `vars`? Also the `role:` in `- role: passwordless_ssh` isn't necessary, but again not sure which cased is considered best practice. Like so:
```
- hosts: passwordless_ssh_servers
  roles:
    - passwordless_ssh
  vars:
    passwordless_ssh_private_key: "{{ lookup('env', 'HOME') }}/mytestkeys/id_rsa"
    passwordless_ssh_public_key: "{{ lookup('env', 'HOME') }}/mytestkeys/id_rsa.pub"
    passwordless_ssh_user: testuser
    passwordless_ssh_user_dir: /home/testuser
```

greg-hellings replied:
The syntax I've used allows me to set specific variables per-role rather than globally setting them for every role included in a playbook. This way, if there are any roles that have name collisions, their variables can be set separate of the other role with the collision.

When setting variables that way, the `role:` part is necessary to differentiate it from the variables.

andrew103 replied:
Makes sense

andrew103 commented on molecule/shared/playbook.yml:
```diff
@@ -0,0 +1,22 @@
+      passwordless_ssh_public_key: "{{ role_path }}/tests/id_rsa.pub"
+      passwordless_ssh_private_key: "{{ role_path }}/tests/id_rsa"
+      passwordless_ssh_user: derp
+      passwordless_ssh_user_home: /home/derp
```
See comment on README

robled commented on molecule/shared/playbook.yml:
```diff
@@ -0,0 +1,22 @@
+      changed_when: false
+      register: ssh_output
+
+    - debug: var=ssh_output.stdout
```
Use indent+colon, ex:

```
- name: Display all variables/facts known for a host
  debug:
    var: hostvars[inventory_hostname]
```

robled commented on molecule/shared/playbook.yml:
```diff
@@ -0,0 +1,22 @@
+      passwordless_ssh_user_home: /home/derp
+  post_tasks: []
+
+- name: test, kinda
```
Make name more descriptive.

andrew103 commented on molecule/shared/playbook.yml:
```diff
@@ -0,0 +1,22 @@
+  hosts: all[0]
+  tasks:
+    - name: try to ssh
+      command: ssh {{ hostvars[groups['all'][1]]['ansible_host'] }} ls
```
Ansible has an [SSH module](https://docs.ansible.com/ansible/2.5/plugins/connection/ssh.html)

greg-hellings replied:
That's a connection plugin, not a module. This is testing that host a and host b from the same group can connect to each other directly, which is not a use case that Ansible directly supports to my knowledge.

robled commented on tasks/main.yml:
```diff
@@ -0,0 +1,38 @@
+    dest: "{{ passwordless_ssh_user_home }}/.ssh/\
+           {{ passwordless_ssh_remote_key_file }}"
+    mode: 0600
+    owner: "{{ passwordless_ssh_user }}"
```
Since this bit is security related it might be a good idea to be explicit about the **group** ownership, as well as the selinux file context, which I believe should be as follows:

`Context: unconfined_u:object_r:ssh_home_t:s0`

greg-hellings replied:
Group is denied permission to read by `mode: 0600`, so I figured the system's default setgid behavior based on users should be OK there. And I don't have any idea what you just said about SELinux.

robled replied:
With selinux enabled, files have extended attributes that can be seen by running `stat ~/.ssh/id_rsa`, for example.  Most ansible modules (such as the copy module) that perform file operations support this stuff, usually with parameters prefixed with "se".  

I do wonder if this operation, without any selinux parameters, will inherit the correct file contexts, or if being explicit is the way to go here.  We could try testing it out and compare what ansible does vs a manually-generated SSH key on a selinux system using the "stat" test.  @seandst knows some selinux stuff, so I'm curious as to what his thoughts are here.

tehsmyers replied:
An option for testing would be to touch the file with the `file` module and see if it just does the right thing with selinux contexts. Based on the `file` module [docs](https://docs.ansible.com/ansible/latest/modules/file_module.html), I believe that if no selinux settings are passed to `file` and if the system supports selinux, the module will use the system defaults for selinux stuff.

The **seuser** arg's doc says "User part of SELinux file context. *Will default to system policy, if applicable.* If set to _default, it will use the user portion of the policy if available." (emphasis mine). Both the **serole** and **seuser** args say that they are set to their respective "...part of SELinux file context, _default feature works as for **seuser**." It's not clear to me that both **serole** and **seuser** will default to the system policy like **seuser** does, but poking in the guts of `ansible.module_utils.basic.AnsibleModule` gives me a little more confidence that just poking the `known_hosts` file before invoking the `known_hosts` modules ought to be good enough (assuming the module works the way you need it to otherwise). Unfortunately, the `known_hosts` module does not invoke the file handling included in `AnsibleModule`, so I think both `file` and `known_hosts` would have to be used to ensure the file is created with the correct selinux context info.

(sudo) `restorecon ~/.ssh/known_hosts` should set the file's selinux contexts to what they should be for easier testing.

robled replied:
I ran a manual test to check the behavior here, and used this file as a reference for proper selinux contexts:

`/etc/selinux/targeted/contexts/files/file_contexts.homedirs`

which gives:

`/home/[^/]+/\.ssh(/.*)? unconfined_u:object_r:ssh_home_t:s0`

I can confirm that the test user `~/.ssh` directory has the correct context set after this role has been used.  The `/root/.ssh` directory has a different default context of `system_u:object_r:ssh_home_t:s0`, which is also handled correctly by this role.

robled commented on tasks/main.yml:
```diff
@@ -0,0 +1,38 @@
+- name: update known hosts with the other hosts in this current group
+  shell: ssh-keyscan {{ hostvars[item]['ansible_host'] |
+         default(hostvars[item]['inventory_hostname']) }} >>
+         {{ passwordless_ssh_user_home }}/.ssh/known_hosts
```
Eventually, someone is going to have a` passwordless_ssh_user_home` with a space in the path, so this should probably get quoted.

robled replied:
Is it possible to use the known_hosts module here?

https://docs.ansible.com/ansible/latest/modules/known_hosts_module.html

greg-hellings replied:
Those people should be defenestrated from high places, but I will add quotes here.

robled replied:
Totally agreed, but you know it's gonna happen.  :dagger:

robled commented on tasks/main.yml:
```diff
@@ -0,0 +1,38 @@
+
+- name: update known hosts with the other hosts in this current group
+  shell: ssh-keyscan {{ item }} >>
+         {{ passwordless_ssh_user_home }}/.ssh/known_hosts
```
Eventually, someone is going to have a` passwordless_ssh_user_home` with a space in the path, so this should probably get quoted.

robled replied:
Is it possible to use the known_hosts module here?

https://docs.ansible.com/ansible/latest/modules/known_hosts_module.html

greg-hellings replied:
*copy, paste*

greg-hellings replied:
Oh. Yeah.

It is.

I was sitting here this morning going, "I wish I didn't have to go write a known_hosts module to make this work right." I just assumed it didn't exist. Don't know why, though.

BRB, completely changing PR.

robled commented on tasks/main.yml:
```diff
@@ -0,0 +1,38 @@
+  register: shell_result
+  changed_when: false
+
+- debug: var=shell_result
```
Was this left in for development?  If not, name the task and use indent+colon syntax.

greg-hellings replied:
This was left in because the source material likes to have this scattered throughout at the end of each such task. Also, I tend to leave names off of debug statements because that's kinda redundant IMO. Is that something we're OK with?

robled replied:
Source material?  It seems that ansible-lint contains a special exception for unnamed debug tasks, so it's probably cool.  I might continue to name my debug tasks just because of habit/consistency.

greg-hellings committed:
Updates from CR

greg-hellings committed:
Another CR change

robled commented on README.md:
````diff
@@ -0,0 +1,77 @@
+```
+- hosts: passwordless_ssh_servers
+  roles:
+    - role: passwordless_ssh
````
Should we go ahead and point this at Ansible Galaxy?

greg-hellings replied:
Done.

robled commented:
This role is missing a **LICENSE** file.

greg-hellings committed:
Update to use `known_hosts` and a few tweaks

Use the known_hosts module, and alter a few code review things to tidy
up the role

greg-hellings committed:
License file added

greg-hellings commented:
`LICENSE` file added. The role should be ready for re-review.

tehsmyers reviewed (approved):
`name: enable ssh, because that's how this all works` :D

andrew103 commented on molecule/shared/playbook.yml:
```diff
@@ -0,0 +1,23 @@
+    - name: try to ssh
+      command: ssh {{ hostvars[groups['all'][1]]['ansible_host'] }} ls -a /
+      become: true
+      become_user: derp
```
Shouldn't this be replaced with `"{{ passwordless_ssh_user }}"` or `"{{ passwordless_ssh_become_user }}"`?

greg-hellings replied:
This is in the molecule test playbook. Those variables are only in scope inside of the role. These are test values I'm using.

andrew103 commented on molecule/shared/prepare.yml:
```diff
@@ -0,0 +1,22 @@
+  tasks:
+    - name: create user derp
+      user:
+        name: derp
```
Shouldn't this be replaced with `"{{ passwordless_ssh_user }}"` or `"{{ passwordless_ssh_become_user }}"`?

greg-hellings replied:
This is setting up the text fixture data and, as such, cannot reference the role's internal parameters.
